File: c++/arrays.cpp
// By: Jeff Listfield //

#include <iostream>
#include <bits/stdc++.h>

int NumberOfTicketsToBuy(int ticket_prices[], int number_of_tickets, int budget)
{
  // We're assuming no negative priced tickets.
  if (ticket_prices == nullptr || budget <= 0)
  {
    return 0;
  }

  // FYI, ticket_prices will be modified by this.
  std::sort(ticket_prices, ticket_prices + number_of_tickets);
  int remaining_budget = budget;
  int tickets_to_buy = 0;
  int ticket_index = 0;
  while (ticket_index < number_of_tickets && ticket_prices[ticket_index] < remaining_budget)
  {
    // Do this before incrementing ticket_index.
    remaining_budget -= ticket_prices[ticket_index];
    tickets_to_buy++;
    ticket_index++;
  }

  return tickets_to_buy;
}

void swap_in_place(int unsorted_array[], int index_a, int index_b)
{
  int temp = unsorted_array[index_a];
  unsorted_array[index_a] = unsorted_array[index_b];
  unsorted_array[index_b] = temp;
}

int NumberOfSwaps(int unsorted_array[], int array_length)
{
  int number_of_swaps = 0;
  int index = 0;
  while (index < array_length)
  {
    if (unsorted_array[index] != index + 1)
    {
      // This moves the item at index to the right place. It does not
      // guarantee that the item moved in to index is in the right spot.
      // We might need to swap this again.
      swap_in_place(unsorted_array, index, unsorted_array[index] - 1);
      number_of_swaps++;
    }
    else
    {
      // Item is in the right spot, move on to the next one.
      index++;
    }
  }
  return number_of_swaps;
}

bool IsPangram(std::string input_string)
{
  // Don't forget to initialize this. (I did.)
  bool character_present[26] = {false};
  int num_characters;
  for (int index = 0; index < input_string.length(); index++)
  {
    char character = input_string[index];
    // Only deal with alphabetical characters.
    if (isalpha(character))
    {
      // Remember to offset this from lowercase 'a'.
      int character_index = tolower(character) - 'a';
      if (!character_present[character_index])
      {
        num_characters++;
        if (num_characters == 26)
        {
          // Once we hit 26 characters, we can return true.
          return true;
        }
      }
      // Make sure we don't double count any character.
      character_present[character_index] = true;
    }
  }
  return false;
}

int main()
{
  int ticket_prices[] = {10, 25, 30, 40};
  int budget = 90;
  int tickets_to_buy = NumberOfTicketsToBuy(ticket_prices, 4, budget);
  std::cout << "You can purchase " << tickets_to_buy << " tickets.\n";

  int unsorted_array[] = {1, 5, 4, 3, 7, 6, 2};
  int number_of_swaps = NumberOfSwaps(unsorted_array, 7);
  std::cout << "Minimum number of swaps is " << number_of_swaps << ".\n";

  std::string string_to_check = "Brown jars prevented the mixture from freezing too quickly";
  bool is_pangram = IsPangram(string_to_check);
  std::cout << "Is the string '" << string_to_check << "' a pangram? " << (is_pangram ? "yes" : "no") << ".\n";
}

File: JavaScript/arrays.js
/**
 * Maximum No. Tickets Purchasable

Given an array of concert ticket prices and a target budget, find the maximum number of tickets that can be purchased without exceeding the target budget.

Example: input -> [10, 25, 30, 40], 90 | output -> 3
 */

function maxTicketsPurchasable(ticketPrices, budget) {
  // taking care of edge cases
  if (ticketPrices.length === 0 || budget <= 0) {
    return 0;
  }

  // let's rule out tickets we already know aren't w/in budget
  let ticketPricesInBudget = [];

  ticketPrices.forEach(price => {
    if (price <= budget) {
      ticketPricesInBudget.push(price);
    }
  });

  // sort tickets so we can pull from the lowest prices upwards
  let sortedTicketPricesInBudget = ticketPricesInBudget.sort((a, b) => a - b);

  // initialize our max ticket count
  let maxTickets = 0;

  // initialize value to track budget - don't modify the input value!
  let budgetRemaining = budget;

  // grab as many tickets as we can until we hit the budget
  // why use a for loop here? gives us the ability to break out of the loop so we don't unecessarily loop through values.
  for (let i = 0; i < sortedTicketPricesInBudget.length; i++) {
    budgetRemaining -= sortedTicketPricesInBudget[i];

    if (budgetRemaining < 0) {
      break;
    }

    maxTickets += 1;
  }

  return maxTickets;
}

// ****************************************

/**
 * Minimum Swaps

Given an unsorted array of consecutive distinct integers, find the minimum number of swaps required to sort the array in ascending order.

Example: input -> [1, 5, 4, 3, 7, 6, 2]  | output -> 3
 */

function swap(i, j, array) {
  // switch values in an array at indices i and j
  let temp = array[i];
  array[i] = array[j];
  array[j] = temp;

  return array;
}

function minSwaps(array) {
  // taking care of edge cases
  if (array.length <= 1) {
    return 0;
  }

  // initialize count value
  let count = 0;

  // initialize array for what will be our sorted values - don't modify the input value!
  // why use the spread operator (...)? We want to create a brand new array of values, not just point to the original array.
  let sortedArray = [...array];

  // go through the array and swap!
  for (let i = 0; i < sortedArray.length; i++) {
    while (sortedArray[i] !== i + 1) {
      sortedArray = swap(i, sortedArray[i] - 1, sortedArray);
      count += 1;
    }
  }

  return count;
}

// ****************************************

/**
 * Pangrams

A pangram is a string that contains every letter of the alphabet. Given a string, write a method that determines if that string is a pangram or not

Example: input -> “Brown jars prevented the mixture from freezing too quickly“  | output -> true
 */

function isPangram(string) {
  // taking care of edge cases
  if (string.length < 26) {
    return false;
  }

  // initialize array with 26 zeroes, each representing a letter in the alphabet
  let array = [...new Array(26)].map(() => 0);

  // initialize return value
  let result = false;

  // go through string's characters and turn array's value to 1 for the associated index (e.g. a = 0, b = 1, etc.)
  for (let i = 0; i < string.length; i++) {
    let character = string.charAt(i).toLowerCase();
    let index = character.charCodeAt(0) - 'a'.charCodeAt(0);

    if (index <= 25 && index >= 0) {
      array[index] = 1;

      if (array.reduce((a, b) => a + b, 0) === 26) {
        result = true;
        break;
      }
    }
  }

  return result;
}
